examples/ar2.py

```diff
@@ -34,41 +34,50 @@
 import os
 import time
 
-import matplotlib
-import matplotlib.pyplot as plt
-import numpy as np
-
 import jax
 from jax import random
 import jax.numpy as jnp
 
 import numpyro
 import numpyro.distributions as dist
-
-matplotlib.use("Agg")
+from numpyro.contrib.control_flow import scan
+from numpyro.handlers import condition
 
 
-def ar2(y, unroll_loop=False):
+def ar2_scan(y):
     alpha_1 = numpyro.sample("alpha_1", dist.Normal(0, 1))
     alpha_2 = numpyro.sample("alpha_2", dist.Normal(0, 1))
     const = numpyro.sample("const", dist.Normal(0, 1))
     sigma = numpyro.sample("sigma", dist.Normal(0, 1))
 
-    def transition_fn(carry, y):
-        y_1, y_2 = carry
-        pred = const + alpha_1 * y_1 + alpha_2 * y_2
-        return (y, y_1), pred
+    def transition(carry, _):
+        y_prev, y_prev_prev = carry
+        m_t = const + alpha_1 * y_prev + alpha_2 * y_prev_prev
+        y_t = numpyro.sample("y", dist.Normal(m_t, sigma))
+        carry = (y_t, y_prev)
+        return carry, None
 
-    if unroll_loop:
-        preds = []
-        for i in range(2, len(y)):
-            preds.append(const + alpha_1 * y[i - 1] + alpha_2 * y[i - 2])
-        preds = jnp.asarray(preds)
-    else:
-        _, preds = jax.lax.scan(transition_fn, (y[1], y[0]), y[2:])
+    timesteps = jnp.arange(y.shape[0] - 2)
+    init = (y[0], y[1])
+ 
+    with numpyro.handlers.condition(data={"y": y[2:]}):
+        scan(transition, init, timesteps)
 
-    mu = numpyro.deterministic("mu", preds)
-    numpyro.sample("obs", dist.Normal(mu, sigma), obs=y[2:])
+
+def ar2_for_loop(y):
+    alpha_1 = numpyro.sample("alpha_1", dist.Normal(0, 1))
+    alpha_2 = numpyro.sample("alpha_2", dist.Normal(0, 1))
+    const = numpyro.sample("const", dist.Normal(0, 1))
+    sigma = numpyro.sample("sigma", dist.Normal(0, 1))
+
+    y_prev = y[0]
+    y_prev_prev = y[1]
```

Comment:
I might not be reading this right, but are they the right way round?

Reply:
Oh, that's right! I called it `y0` and `y1` when I wrote my version but I was thinking backwards in time, so `y1` should be older than `y0`.

Reply:
good catch!